File: frontend/src/types/project.ts
export const ROLES: Record<string, TaskRole> = {
  OWNER: "Owner",
  MEMBER: "Member",
  ASSIGNEE: "Assignee",
};

export type TaskRole = "Owner" | "Member" | "Assignee";

export type ProjectRole = "Owner" | "Member";

export type OwnerApi = {
  id: string;
  username: string;
  display_name: string;
};

export type Owner = {
  id: string;
  username: string;
  displayName: string;
};

export interface ProjectSummaryApi {
  id: string;
  name: string;
  description?: string;
  skills?: string;
  role: ProjectRole;
  unassigned_tasks: number;
  ongoing_tasks: number;
  completed_tasks: number;
  created_at: string;
  updated_at?: string;
}

export interface ProjectSummary {
  id: string;
  name: string;
  description?: string;
  skills?: string;
  role: ProjectRole;
  unassignedTasks: number;
  ongoingTasks: number;
  completedTasks: number;
  createdAt: string;
  updatedAt?: string;
}

export interface ProjectsApiResponse {
  projects: ProjectSummaryApi[];
  page: number;
  limit: number;
  hasNext: boolean;
}

export const MapProject = (p: ProjectSummaryApi): ProjectSummary => ({
  id: p.id,
  name: p.name,
  description: p.description,
  skills: p.skills,
  role: p.role,
  unassignedTasks: p.unassigned_tasks,
  ongoingTasks: p.ongoing_tasks,
  completedTasks: p.completed_tasks,
  createdAt: p.created_at,
  updatedAt: p.updated_at,
});

export interface ProjectDetailsApi {
  id: string;
  name: string;
  description?: string;
  skills?: string;
  role: ProjectRole;
  owner: OwnerApi;
  unassigned_tasks: number;
  ongoing_tasks: number;
  completed_tasks: number;
  abandoned_tasks: number;
  members_count: number;
  created_at: string;
  updated_at?: string;
}

export interface ProjectDetails {
  id: string;
  name: string;
  description?: string;
  skills?: string;
  role: ProjectRole;
  owner: Owner;
  unassignedTasks: number;
  ongoingTasks: number;
  completedTasks: number;
  abandonedTasks: number;
  membersCount: number;
  createdAt: string;
  updatedAt?: string;
}

export const MapProjectDetails = (p: ProjectDetailsApi): ProjectDetails => ({
  id: p.id,
  name: p.name,
  description: p.description,
  skills: p.skills,
  owner: {
    id: p.owner.id,
    username: p.owner.username,
    displayName: p.owner.display_name,
  },
  role: p.role,
  unassignedTasks: p.unassigned_tasks,
  ongoingTasks: p.ongoing_tasks,
  completedTasks: p.completed_tasks,
  abandonedTasks: p.abandoned_tasks,
  membersCount: p.members_count,
  createdAt: p.created_at,
  updatedAt: p.updated_at,
});

type JoinStatus = "Pending" | "Accepted" | "Rejected";

export interface JoinRequest {
  projectId: string;
  userId: string;
  username: string;
  displayName: string;
  email: string;
  avatarURL: string;
  isActive: boolean;
  status: JoinStatus;
  createdAt: string;
}

export interface JoinRequestApi {
  project_id: string;
  user_id: string;
  username: string;
  display_name: string;
  email: string;
  avatar_url: string;
  is_active: boolean;
  status: JoinStatus;
  created_at: string;
}

export const MapJoinRequest = (request: JoinRequestApi): JoinRequest => ({
  projectId: request.project_id,
  userId: request.user_id,
  username: request.username,
  displayName: request.display_name,
  email: request.email,
  avatarURL: request.avatar_url,
  isActive: request.is_active,
  status: request.status,
  createdAt: request.created_at,
});

export interface JoinRequestsResponseApi {
  join_requests: JoinRequestApi[];
}

export interface Member {
  projectId: string;
  id: string;
  username: string;
  displayName: string;
  email: string;
  avatarUrl: string;
  isActive: boolean;
  role: ProjectRole;
  createdAt: string;
  updatedAt: string;
}

export interface MemberApi {
  project_id: string;
  id: string;
  username: string;
  display_name: string;
  email: string;
  avatar_url: string;
  is_active: boolean;
  role: ProjectRole;
  created_at: string;
  updated_at: string;
}

export interface MembersResponse {
  members: MemberApi[];
}

export const MapMember = (m: MemberApi): Member => ({
  projectId: m.project_id,
  id: m.id,
  username: m.username,
  displayName: m.display_name,
  email: m.email,
  avatarUrl: m.avatar_url,
  isActive: m.is_active,
  role: m.role,
  createdAt: m.created_at,
  updatedAt: m.updated_at,
});

File: backend/services/project.go
package services

import (
	"context"
	"fmt"
	"slices"
	"strings"

	"github.com/ptracker/apierr"
	"github.com/ptracker/domain"
	"github.com/ptracker/interfaces"
)

type projectService struct {
	store             interfaces.Store
	projectPermission *ProjectPermissionService
}

func NewProjectService(store interfaces.Store) *projectService {
	permissionService := &ProjectPermissionService{
		store: store,
	}
	return &projectService{
		store:             store,
		projectPermission: permissionService,
	}
}

func (s *projectService) CreateProject(ctx context.Context, name string,
	description, skills *string,
	owner string) (string, error) {
	var projectId string

	if strings.Trim(name, " ") == "" {
		return "", apierr.ErrInvalidValue
	}

	err := s.store.WithTx(ctx, func(txStore interfaces.Store) error {
		var err error

		projectId, err = txStore.Project().Create(ctx, name, description, skills, owner)
		if err != nil {
			return fmt.Errorf("store project create: %w", err)
		}

		err = txStore.Role().Create(ctx, projectId, owner, domain.ROLE_OWNER)
		if err != nil {
			return fmt.Errorf("store role create: %w", err)
		}

		return nil
	})
	if err != nil {
		return "", fmt.Errorf("store WithTx: %w", err)
	}

	return projectId, nil
}

func (s *projectService) ListProjects(ctx context.Context, userId string) ([]*domain.PrivateProjectListed, error) {
	projects, err := s.store.List().PrivateProjects(ctx, userId)
	if err != nil {
		return projects, fmt.Errorf("store project all: %w", err)
	}

	return projects, nil
}

func (s *projectService) GetPrivateProject(ctx context.Context,
	projectId, userId string) (*domain.ProjectDetail, error) {

	permitted, err := s.projectPermission.CanAccess(ctx, projectId, userId)
	if err != nil {
		return nil, fmt.Errorf("permission service can access: %w", err)
	}

	if !permitted {
		return nil, apierr.ErrForbidden
	}

	project, err := s.store.Project().Get(ctx, projectId)
	if err != nil {
		return nil, fmt.Errorf("store project all: %w", err)
	}

	userRole, err := s.store.Role().Get(ctx, projectId, userId)
	if err != nil {
		return nil, fmt.Errorf("store role get: %w", err)
	}

	owner, err := s.store.User().Get(ctx, userId)
	if err != nil {
		return nil, fmt.Errorf("store user get: %w", err)
	}

	memberCount, err := s.store.Role().CountMembers(ctx, projectId)
	if err != nil {
		return nil, fmt.Errorf("store role count members: %w", err)
	}

	return &domain.ProjectDetail{
		ProjectSummary: &domain.ProjectSummary{
			Project: &domain.Project{
				Id:          project.Id,
				Name:        project.Name,
				Description: project.Description,
				Skills:      project.Skills,
				CreatedAt:   project.CreatedAt,
				UpdatedAt:   project.UpdatedAt,
			},
			UnassignedTasks: project.UnassignedTasks,
			OngoingTasks:    project.OngoingTasks,
			CompletedTasks:  project.CompletedTasks,
			AbandonedTasks:  project.AbandonedTasks,
		},
		Role:        userRole.Role,
		MemberCount: memberCount,
		Owner: &domain.Member{
			Id:          owner.Id,
			Username:    owner.Username,
			DisplayName: owner.DisplayName,
			Email:       owner.Email,
			AvatarURL:   owner.AvatarURL,
			IsActive:    owner.IsActive,
			CreatedAt:   userRole.CreatedAt,
			UpdatedAt:   userRole.UpdatedAt,
		},
	}, nil
}

func (s *projectService) GetProjectMembers(ctx context.Context,
	projectId, userId string) ([]*domain.Member, error) {

	permitted, err := s.projectPermission.CanSeeMembers(ctx, projectId, userId)
	if err != nil {
		return nil, fmt.Errorf("permission service can access: %w", err)
	}

	if !permitted {
		return nil, apierr.ErrForbidden
	}

	members, err := s.store.List().Members(ctx, projectId)
	if err != nil {
		return nil, fmt.Errorf("store list members: %w", err)
	}

	return members, nil
}

func (s *projectService) ListJoinRequests(ctx context.Context,
	projectId, userId string) ([]*domain.JoinRequestListed, error) {

	permitted, err := s.projectPermission.CanSeeMembers(ctx, projectId, userId)
	if err != nil {
		return nil, fmt.Errorf("permission service can access: %w", err)
	}

	if !permitted {
		return nil, apierr.ErrForbidden
	}

	joinRequests, err := s.store.List().JoinRequests(ctx, projectId)
	if err != nil {
		return nil, fmt.Errorf("store list join requests: %w", err)
	}

	return joinRequests, nil
}

func (s *projectService) AccessOrRejectJoinRequest(ctx context.Context,
	projectId, ownerId, requestorId, joinStatus string) error {

	permitted, err := s.projectPermission.CanRespondToJoinRequests(ctx, projectId, ownerId)
	if err != nil {
		return fmt.Errorf("permission service can access: %w", err)
	}
	if !permitted {
		return apierr.ErrForbidden
	}

	if !slices.Contains([]string{
		domain.JOIN_STATUS_PENDING,
		domain.JOIN_STATUS_ACCEPTED,
		domain.JOIN_STATUS_REJECTED,
	}, joinStatus) {
		return apierr.ErrInvalidValue
	}

	userRole := domain.ROLE_MEMBER

	status, err := s.store.JoinRequest().Get(ctx, projectId, requestorId)
	if err != nil {
		return fmt.Errorf("store join request get: %w", err)
	}

	if status == joinStatus {
		return apierr.ErrInvalidValue
	}

	// Y Rejected -> Pending
	// X Rejected -> Accepted
	// X Accepted -> Pending|Rejected

	if (status == domain.JOIN_STATUS_REJECTED &&
		slices.Contains([]string{
			domain.JOIN_STATUS_ACCEPTED,
		}, joinStatus)) ||
		(status == domain.JOIN_STATUS_ACCEPTED &&
			slices.Contains([]string{
				domain.JOIN_STATUS_PENDING,
				domain.JOIN_STATUS_REJECTED,
			}, joinStatus)) {
		return apierr.ErrInvalidValue
	}

	err = s.store.WithTx(ctx, func(txStore interfaces.Store) error {
		var err error

		err = txStore.JoinRequest().Update(ctx, projectId, requestorId, joinStatus)
		if err != nil {
			return fmt.Errorf("store join request update: %w", err)
		}

		if joinStatus == domain.JOIN_STATUS_ACCEPTED {
			err = txStore.Role().Create(ctx, projectId, requestorId, userRole)
			if err != nil {
				return fmt.Errorf("store role create: %w", err)
			}
		}

		return nil
	})
	if err != nil {
		return fmt.Errorf("transaction: %w", err)
	}

	return nil
}
